keps/sig-apps/2879-ready-pods-job-status/README.md

```diff
@@ -0,0 +1,319 @@
+- [x] (R) KEP approvers have approved the KEP status as `implementable`
+- [x] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests for meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [x] (R) Production readiness review completed
+- [x] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The Job status has a field `active` which counts the number of Job Pods that
+are in `Running` or `Pending` phases. In this KEP, we add a field `ready` that
+counts the number of Job Pods that have a `Ready` condition, with the same
+best effort guarantees as the existing `active` field.
+
+## Motivation
+
+Job Pods can remain in the `Pending` phase for a long time in clusters with
+tight resources and when image pulls take long. Since the `Job.status.active`
+field includes `Pending` Pods, this can give a false impression of progress
+to end users or other controllers. This is more important when the pods serve
+as workers and need to communicate among themselves.
+
+A separate `Job.status.ready` field can provide more information for users
+and controllers, reducing the need to listen to Pod updates themselves.
+
+Note that other workload APIs (such as ReplicaSet and StatefulSet) have a
+similar field: `.status.readyReplicas`.
+
+### Goals
+
+- Add the field `Job.status.ready` that keeps a count of Job Pods with the
+  `Ready` condition.
+
+### Non-Goals
+
+- Provide strong guarantees for the accuracy of the count. Due to the
+  asynchronous nature of k8s, there are can be more or less Pods currently
+  ready than what the count provides.
+
+## Proposal
+
+Add the field `.status.ready` to the Job API. The job controller updates the
+field based on the number of Pods that have the `Ready` condition.
```

Comment:
Should we consider some kind of batching (e.g. no more than 1 call that updates only Ready field per X seconds per Job)?

I'm afraid that higher granularity will not provide any value (given the informative nature of a field, it pretty much should be consumed only be humans), and much higher load on kube-apiserver may cause issues.

Reply:
I agree that full granularity might not be very useful, but knowing if 100% of the parallelism has been reached is probably the most important thing for users. So whatever we do to batch, we should special case 100% to always go through.

Two thoughts:
- Why are you worried about the load from Jobs but not about existing `ready` fields in other workload APIs?
- An alternative would be to batch based on a percentage: only update the field if other fields changed or it changed more than X% (X=25?)

Reply:
> Why are you worried about the load from Jobs but not about existing ready fields in other workload APIs?

I'm less worried, because generally batch oriented workloads are much more short-living, so those changes much be much more frequent. 

> An alternative would be to batch based on a percentage: only update the field if other fields changed or it changed more than X% (X=25?)

X=25% sounds a bit high for large enough jobs. But yeah, that's an option. I don't have very strong opinion on what exact batching to use, but I think we need *some*.

Reply:
I dont understand this

Reply:
Added in `Risks and Mitigations`.

Reply:
I think I'm inclined to agree with Wojtek here, when we create pods we'll update the job status once per cycle which can create from 1 to parallelism of pods. Having a similar mechanism for ready field would be beneficial api-call wise. But I'm ok pushing this to beta stage.

Reply:
What do you mean by similar mechanism? For creations, the job-controller knows how many pods it is creating, so it can set up the field easily.
But for ready pods, the job-controller has zero control. What do you think of the batch proposal I presented?

Reply:
I suggested changes to the approach.

Reply:
Yeah, maybe similar is a wrong word, what I meant is having a batching of some sort :wink:

Reply:
Ack